### PR TITLE
Add filter around fetch link suggestions

### PR DESCRIPTION
**Pull request by Aljullu**
## Description
This PR adds a filter around the suggestions returned by `__experimental-fetch-link-suggestions.js`. This will allow plugins to add additional custom suggestions to the URL input.

Use case: several plugins have custom endpoints and they want to make it easy for users to add links to those endpoints. For example, in WooCommerce we would like to make it easy to link to some endpoint like _Orders_, _Payment methods_, _Log out_...

That's specially important for the Navigation editor/block. In the past, there were hooks for that (#31316), but those hooks are not currently honored by the Navigation block and the PR which added support for them didn't have any movement in the last four months (#31584). So I'm trying a different approach here. :slightly_smiling_face: 

### Testing steps

- Open the Navigation editor.
- Open your browser console and run one of these two code snippets (this is what plugins would do on their end):
<details>
<summary>Sync example</summary>

```JS
wp.hooks.addFilter(
	'editor.fetchLink.suggestions',
	'wikipedia_link_suggestion',
	( results, search ) => {
		const wikiLinks = [
			{
				id: 'wikipedia',
				url: 'https://www.wikipedia.org/',
				title: 'Wikipedia',
				type: 'Wikipedia link',
			},
			{
				id: 'wikipedia_en',
				url: 'https://en.wikipedia.org/',
				title: 'Wikipedia (English)',
				type: 'Wikipedia link',
			},
		].filter( ( link ) =>
			link.title.toLowerCase().includes( search.toLowerCase() )
		);

		return results.concat( wikiLinks );
	}
);
```

</details>
<details>
<summary>Async example</summary>

```JS
wp.hooks.addFilter(
	'editor.fetchLink.suggestions',
	'wikipedia_search_suggestion',
	( results, search ) => {
		return new Promise( ( resolve ) => {
			setTimeout( () => {
				const wikiLinks = [
					{
						id: 'wikipedia',
						url: 'https://www.wikipedia.org/',
						title: 'Wikipedia',
						type: 'custom_link',
					},
					{
						id: 'wikipedia_en',
						url: 'https://en.wikipedia.org/',
						title: 'Wikipedia (English)',
						type: 'custom_link',
					},
				].filter( ( link ) =>
					link.title.toLowerCase().includes( search.toLowerCase() )
				);
				resolve( results.concat( wikiLinks ) );
			}, 300 );
		} );
	}
);
```

</details>

- Add a link inside the Navigation.
- Start typing `wiki`.
- Verify the `Wikipedia` and `Wikipedia (English)` suggestions appear.
- Verify you can add those links, save the page and the links are persisted.
- Repeat the process above inside the post/page editor using the Navigation block.

## Screenshots

https://user-images.githubusercontent.com/3616980/138442146-b2057162-6cec-410c-b3b8-a3a1cbc09afc.mp4

## Types of changes
Introduces a new JS filter: `editor.fetchLink.suggestions`.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by Aljullu** [comment]
@getdave @jasmussen wondering if you would have time to take a look at this or maybe find somebody else who can? I'm happy to discuss alternative suggestions as well if you have some in mind.

**Comment by jasmussen** [comment]
Hey, thanks for the PR! Dave would be a good one to chime in on this one as well.

In [this comment](https://github.com/WordPress/gutenberg/issues/30170#issuecomment-920719539), we discuss the shortcomings of surfacing the most recent 3 published items and then falling back to search, and it contains some mockups that could surface content in a more compact way from various sources. I understand that's a relatively big endeavor, so not a 5.9 thing. But would such a mechanism be an alternative to, or addition, to what's being proposed here?

File: packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js
Comment (getdave) [inline packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js]: What happens if - for example - I wanted to asynchronously retrieve additional results here from an API endpoint? If you wrap the results of the hook in `Promise.resolve()` then it will handle both promises and non-promises.

**Comment by Aljullu** [comment]
> But would such a mechanism be an alternative to, or addition, to what's being proposed here?

Thanks for sharing those mock-ups @jasmussen. I expect they will work well for our needs. How do you envision extensions will hook into the list of available suggested links there? Registering a new Navigation Link block?

**Review by getdave — COMMENTED** [review]
Thanks for the PR @Aljullu.

Overall this seems like a reasonable addition to me. I have a question regarding providing results async.

**Comment by jasmussen** [comment]
> How do you envision extensions will hook into the list of available suggested links there? Registering a new Navigation Link block?

I should reiterate that particular design appears unlikely to land with 5.9. That also means the precise way it works is not fully hashed out, and I'd always defer to a developer as I'm not fully qualified to speak to the innards. But yes, it does seem at the face of it that the particular way to filter based on post type could come from the navigation item type. 

This also speaks to the need for better transform tools, as per #34041 you'll always just insert Page Links by default, at least in the Navigation block.

File: packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js
Comment (Aljullu) [reply packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js]: Thanks for the review @getdave, I updated the code in 1baedc2aa97d8a827c98c0123e800197ba04bd00, let me know if that's what you had in mind. I also updated the PR description with an async code snippet so it can be tested.

**Review by getdave — CHANGES_REQUESTED** [review]
Let's also augment the test suite to include tests for this new filter.

Also wondering whether the filter name itself should be in an "experimental" namespace. Are there docs or conventions for that... 🤔 

Lastly, I'd like to get a second opinion here from the Core team. I know there has previously been push back on adding Hooks. I should note that we appear to already have [a similar filter in the `AutoComplete` component](https://github.com/WordPress/gutenberg/blob/e99df03cd855b6b64ff03d38f4d2b3ca0e57e5c5/packages/block-editor/src/components/autocomplete/index.js#L51).

**Comment by gwwar** [comment]
🤔 Have we considered allowing a full replacement of the component instead, perhaps like MediaPlaceholder? Folks can then use whatever custom endpoints they'd prefer to control the flow after clicking on the inserter.

I'd be very hesitant to add a hook to something that is still experimental and one that we're still exploring UX options on.

**Comment by Aljullu** [comment]
Thanks for following up and pinging other reviewers @getdave! For now, I added tests in 06f306a94f1a8f36d3ba56355766423e8a534415 and prefixed the filter with `experimental` in 59ad6fc422370eddfcc1ffbbdd765f28cf34a3b8.

> 🤔 Have we considered allowing a full replacement of the component instead, perhaps like MediaPlaceholder? Folks can then use whatever custom endpoints they'd prefer to control the flow after clicking on the inserter.

Not sure if that question is for me. :slightly_smiling_face: But if I'm understanding it right, you mean plugins would create their own `LinkControl` component to replace the Gutenberg one on certain Navigation Link variations, is that what you meant?

**Comment by jasmussen** [comment]
> Have we considered allowing a full replacement of the component instead, perhaps like MediaPlaceholder? Folks can then use whatever custom endpoints they'd prefer to control the flow after clicking on the inserter.

I like the sound of that. This might also allow us to try out a small page list tree view component just for the Page Link menu item type, without doing the full refactor?

**Comment by getdave** [comment]
> I'd be very hesitant to add a hook to something that is still experimental and one that we're still exploring UX options on.

This reflects my concern. I would say however, that it's asking a lot for each extender to re-implement LinkControl.

I think the bottom line is that LinkControl needs work to become more flexible. This PR would be a sticking plaster over the underlying problem.

FYI I'm looking into a few option for flexibility now but in the short term perhaps Kerry's suggestion is best.

**Comment by Aljullu** [comment]
I'm going to close this PR as it's almost two years old and I'm not actively working on it.

We still have an issue in the WC Blocks repo regarding this https://github.com/woocommerce/woocommerce/issues/42670, so we might explore alternative solutions in the future if this gets prioritized. Thanks everybody who participated in the discussions!

**Pull request closed without merging**
